lib/ansible/executor/task_executor.py

```diff
@@ -269,28 +269,39 @@ def _squash_items(self, items, variables):
             if all(isinstance(o, string_types) for o in items):
                 final_items = []
                 name = self._task.args.pop('name', None) or self._task.args.pop('pkg', None)
-                # The user is doing an upgrade or some other operation
-                # that doesn't take name or pkg.
-                if name:
+
+                # This gets the information to check whether the name field
+                # contains a template that we can squash for
+                template_no_item = template_with_item = None
+                if templar._contains_vars(name):
+                    variables['item'] = '\0$'
+                    template_no_item = templar.template(name, variables, cache=False)
+                    variables['item'] = '\0@'
+                    template_with_item = templar.template(name, variables, cache=False)
+                    del variables['item']
+
+                # Check if the user is doing some operation that doesn't take
+                # name/pkg or the name/pkg field doesn't have any variables
+                # and thus the items can't be squashed
+                if name and (template_no_item != template_with_item):
                     for item in items:
                         variables['item'] = item
                         if self._task.evaluate_conditional(templar, variables):
```

Comment:
The _task.evaluate_conditional() call is so that users can filter some of the items in their playbook.  We're not sure if it's really a valuable feature but theoretically one cold do something like:
```
- yum: name="{{item}}"
  with_items: ["yum", "dnf"]
  when: "'yum' in item and distribution_family == 'RedHat'"
```
to take advantage of this.